src/module_3/assign_3.md

```diff
@@ -0,0 +1,980 @@
+df_orders = df.groupby(by='order_id')['outcome'].sum()
+print(len(df_orders))
+transac_5_items = df_orders[df_orders>5].index.to_list()
+```
+
+    3446
+
+
+
+```python
+#Splitting for test/train/validation, keeping the transactions in the same dataset to avoid data leakeage
+
+random.shuffle(transac_5_items)
+
+train_data_orders =  transac_5_items[:int(len(transac_5_items)*0.7)]
+val_data_orders = transac_5_items[int(len(transac_5_items)*0.7):int(len(transac_5_items)*0.9)]
+test_data_orders = transac_5_items[int(len(transac_5_items)*0.9):]
+```
+
+### Main Models
+
+
+```python
+train_data_x = df[df['order_id'].isin(train_data_orders)].iloc[:,4:]
+train_data_y = df[df['order_id'].isin(train_data_orders)].iloc[:,0]
+val_data_x = df[df['order_id'].isin(val_data_orders)].iloc[:,4:]
+val_data_y = df[df['order_id'].isin(val_data_orders)].iloc[:,0]
+test_data_x = df[df['order_id'].isin(test_data_orders)].iloc[:,4:]
+test_data_y = df[df['order_id'].isin(test_data_orders)].iloc[:,0]
+```
+
+
+```python
+def draw_roc_prc(real_y, y_prob, title):
+    '''
+    Plots a ROC and a Precision and Recall curve.
+    '''
+    fig, ax = plt.subplots(1, 2, figsize=(10, 4))
+    fpr, tpr, thresholds = metrics.roc_curve(real_y, y_prob)
+    roc_auc = metrics.auc(fpr, tpr)
+    
+    fig.suptitle(f'ROC and Precision-Recall Curves: {title}')
+
+    ax[0].set_title(f'ROC curve (AUC = {roc_auc:.2f})')
+    ax[0].plot(fpr, tpr, color='darkorange', lw=2)
+    ax[0].plot([0, 1], [0, 1], color='navy', lw=2, linestyle='--')
+    ax[0].set_xlim([0.0, 1.0])
+    ax[0].set_ylim([0.0, 1.05])
+    ax[0].set_xlabel('False Positive Rate')
+    ax[0].set_ylabel('True Positive Rate')
+
+    precision, recall, thresholds = metrics.precision_recall_curve(real_y, y_prob)
+
+    average_precision = metrics.average_precision_score(real_y, y_prob)
+
+    ax[1].plot(recall, precision, marker='.')
+    ax[1].set_xlabel('Recall')
+    ax[1].set_ylabel('Precision')
+    ax[1].set_title('Precision-Recall Curve (AUC-PR = {:.2f})'.format(average_precision))
+    plt.tight_layout()
+    plt.show()
+
+
+```
+
+
+```python
+#Base model with no penalties
+log_model_base = LogisticRegression(penalty= None, n_jobs=-1, class_weight='balanced').fit(train_data_x,train_data_y)
+
+y_train_pred_base = log_model_base.predict(train_data_x)
+y_train_prob_base = log_model_base.predict_proba(train_data_x)[:,1]
+
+draw_roc_prc(train_data_y, y_train_prob_base,"No regularization")
+print(metrics.confusion_matrix(train_data_y,y_train_pred_base))
+
+```
+
+
+    
+![png](assign_3_files/assign_3_12_0.png)
+    
+
+
+    [[1394040       0]
+     [  21434       0]]
+
+
+
+```python
+def best_threshold(y_true:pd.core.series.Series, y_prob:np.ndarray)->int:
+    '''
+    Function to obtain the threshold that minimizes false positive to be the closest possible to 1%.
+    '''
+    mean = y_prob.mean()
+    std = y_prob.std()
+    thresholds = np.arange(mean - 3*std, mean + 3*std, std/5)
```

Comment:
Why? Is it normally distributed? What if your model produces a bimodal distribution with all the probability mass around 0 for negative class and around 1 for positive class?

Reply:
I didn't think about this, good point. Anyways, I change the way I use it to not fall into this trap.